launcher/ui/widgets/WideBar.cpp

```diff
@@ -23,137 +28,237 @@ private slots:
+{
     auto iter = getMatching(after);
-    if(iter == m_entries.end())
+    if (iter == m_entries.end())
         return;
 
-    auto entry = new BarEntry();
-    entry->qAction = insertWidget((*(iter+1))->qAction, new ActionButton(action, this));
-    entry->wideAction = action;
-    entry->type = BarEntry::Action;
+    BarEntry entry;
+    entry.bar_action = insertWidget((iter + 1)->bar_action, new ActionButton(action, this));
+    entry.menu_action = action;
+    entry.type = BarEntry::Type::Action;
+
     m_entries.insert(iter + 1, entry);
+
+    m_menu_state = MenuState::Dirty;
 }
 
 void WideBar::insertSpacer(QAction* action)
 {
     auto iter = getMatching(action);
-    if(iter == m_entries.end())
+    if (iter == m_entries.end())
         return;
 
-    QWidget* spacer = new QWidget();
+    auto* spacer = new QWidget();
     spacer->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
 
-    auto entry = new BarEntry();
-    entry->qAction = insertWidget((*iter)->qAction, spacer);
-    entry->type = BarEntry::Spacer;
+    BarEntry entry;
+    entry.bar_action = insertWidget(iter->bar_action, spacer);
+    entry.type = BarEntry::Type::Spacer;
     m_entries.insert(iter, entry);
 }
 
 void WideBar::insertSeparator(QAction* before)
 {
     auto iter = getMatching(before);
-    if(iter == m_entries.end())
+    if (iter == m_entries.end())
         return;
 
-    auto entry = new BarEntry();
-    entry->qAction = QToolBar::insertSeparator(before);
-    entry->type = BarEntry::Separator;
+    BarEntry entry;
+    entry.bar_action = QToolBar::insertSeparator(before);
+    entry.type = BarEntry::Type::Separator;
+
     m_entries.insert(iter, entry);
 }
 
-QMenu * WideBar::createContextMenu(QWidget *parent, const QString & title)
+QMenu* WideBar::createContextMenu(QWidget* parent, const QString& title)
 {
-    QMenu *contextMenu = new QMenu(title, parent);
-    for(auto & item: m_entries) {
-        switch(item->type) {
+    auto* contextMenu = new QMenu(title, parent);
+    for (auto& item : m_entries) {
+        switch (item.type) {
             default:
-            case BarEntry::None:
+            case BarEntry::Type::None:
                 break;
-            case BarEntry::Separator:
-            case BarEntry::Spacer:
+            case BarEntry::Type::Separator:
+            case BarEntry::Type::Spacer:
                 contextMenu->addSeparator();
                 break;
-            case BarEntry::Action:
-                contextMenu->addAction(item->wideAction);
+            case BarEntry::Type::Action:
+                contextMenu->addAction(item.menu_action);
                 break;
         }
     }
     return contextMenu;
 }
 
+static void copyAction(QAction* from, QAction* to)
+{
+    Q_ASSERT(from);
+    Q_ASSERT(to);
+
+    to->setText(from->text());
+    to->setIcon(from->icon());
+    to->setToolTip(from->toolTip());
+}
+
+void WideBar::contextMenuEvent(QContextMenuEvent* event)
+{
```

Comment:
maybe use [`QWidget::customContextMenuRequested` signal](https://doc.qt.io/qt-6/qwidget.html#customContextMenuRequested) instead? That's what we use for most of the codebase.